src/compiler/scala/tools/nsc/typechecker/Implicits.scala

```diff
@@ -1581,4 +1591,8 @@ trait ImplicitsStats {
   val matchesPtNanos      = newSubTimer  ("  matchesPT", typerNanos)
   val implicitCacheAccs   = newCounter   ("implicit cache accesses", "typer")
   val implicitCacheHits   = newSubCounter("implicit cache hits", implicitCacheAccs)
+
+  import scala.reflect.internal.util.Position
+  final val implicitSearchesByType = symbolTable.perRunCaches.newMap[global.Type, Int]()
+  final val implicitSearchesByPos = symbolTable.perRunCaches.newMap[Position, Int]()
```

Comment:
Instead of collecting this data within the compiler (this data is collected when ever statistics are enabled, but only used by your compiler plugin), I'd prefer some mechanism to allow a compiler plugin to listen for certain events. See [`AnalyzerPlugins`](https://github.com/scala/scala/blob/v2.12.3/src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala#L281-L288) for an example. Here it would be simpler than with analyzer plugins, because it's only about listening for notifications and not changing the result of any method call.

I'm open for other suggestions.

Reply:
This was not addressed. The goal here is to provide a hook for compiler plugins, so I really feel that it's not great to collect some very specific data inside the compiler itself.

Reply:
Hey Lukas, I don't have a clear idea how we could engineer a general interface to listen for general events in the compiler that adds more value than this current implementation.
  
The problem I see here is that these hooks are very explicit and require instrumentation per ever kind of event (in different parts of the compiler).
  
In `AnalyzerPlugins`, I see a more general interface that allows plugins to refine type signatures in certain places, giving a feeling that analyzer plugins can complement typechecking. I don't see a way we can do this for explicit event listening because this requires:
  
* The creation of the timers and subtimers.
* Exposing these timers and subtimers explicitly in the `statistics` field, so that they are accessible from any call-site of the compiler.
* Instrumenting one call-site per ever event you want to listen.
  
Maybe I'm missing something?

Reply:
I think that implicit search should be instrumented here in the compiler to keep track of some important statistics and aggregate these into `SearchResult`. The most important stat is the overall number of `typedImplicit1` calls made within an implicit search tree. Perhaps the number of implicit macros expanded is also interesting. We should also have an easy way for the tools which process the data to distinguish "top level" implicit searches from recursive ones.

A new extension point in `AnalyzerPlugin`-s could be added to listen on the result of `inferImplicit`, and then the work to slice, dice, and present this data could be left in the hands of the plugin.

Reply:
Thanks for the input. If this PR can still make it for 2.12.4 I'm happy to have a minimal solution here for now and implement a more structured approach later. Would that work for you, @retronym? Would you be OK with the two hash maps?

Reply:
@retronym @lrytz This sounds like a good idea. I'm implementing it real quick. However, I'll add two hooks into implicit search, one before the search and the other after the search. This will help me add timers and not only counters.

Reply:
Implemented in https://github.com/scala/scala/pull/6067/commits/41d60f836be54a7010c3656e5e28107451bfea54.